Add IQS021 rule (#2386)
home-assistant/developers.home-assistant: docs/core/integration-quality-scale/rules/iqs021.md

---
title: "IQS021 - Clean up stale devices"
---

IQS021 is as follows:

Clean up stale devices

## Reasoning

When a device is removed from a hub or account, it should also be removed from Home Assistant.
This way, the user interface will not show devices that are no longer available.

> We should only remove devices that we are sure are no longer available.

Author: I think integrations that can't say this for sure should implement the `async_remove_config_entry_device` so users can delete them themselves as a last resort. Should this be in the rule or should this be in the exceptions

Reviewer: I'd put it in the rule.

This way, the user interface will not show devices that are no longer available.

We should only remove devices that we are sure are no longer available.

## Example implementation

In this example, we have a coordinator that fetches data from a service.
When the data is updated, we check if any devices have been removed.
If so, we remove them from the device registry.
This also causes all entities associated with the device to be removed.

`coordinator.py`
```python
class MyCoordinator(DataUpdateCoordinator[dict[str, MyDevice]]):
    """Class to manage fetching data."""

    def __init__(self, hass: HomeAssistant, client: MyClient) -> None:
        """Initialize coordinator."""
        super().__init__(
            hass,
            logger=LOGGER,
            name=DOMAIN,
            update_interval=timedelta(minutes=1),
        )
        self.client = client
        self.previous_devices: set[str] = set()

    async def _async_update_data(self) -> dict[str, MyDevice]:
        try:
            data = await self.client.get_data()
        except MyException as ex:
            raise UpdateFailed(f"The service is unavailable: {ex}")

>         current_devices = set(data.keys())

Reviewer: ```suggestion
        current_devices = set(data)
```

>         if (old_devices := self.previous_devices - current_devices):

Reviewer: ```suggestion
        if (stale_devices := self.previous_devices - current_devices):
```

            device_registry = dr.async_get(self.hass)

>             for device_id in old_devices:

Reviewer: ```suggestion
            for device_id in stale_devices:
```

This way, the user interface will not show devices that are no longer available.

We should only remove devices that we are sure are no longer available.
If you can't be sure if a device is still available, be sure to implement `async_remove_config_entry_device`.

> This allows users to delete the device from the device registry manually.

Reviewer: ```suggestion
This allows the user to delete the device from the device registry manually.
```

        return data
```

To show a second example where someone can delete the device from the device registry manually, we implement `async_remove_config_entry_device` in `__init__.py`.

> Having this function defined will enable the delete button in the dialog.

Reviewer: ```suggestion
Having this function defined will enable the delete button on the device page in the UI.
```
